### ISSUE TITLE
tf.keras.layers.ReLU() drops tensor._keras_mask (sample_weight) unintentionally.

### ISSUE THREAD

**Issue by Rayarrow** [tensorflow/tensorflow#36894]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information** 
Windows 10 1903, TF 2.0 or 2.1, Python 3.7, CUDA10.1, cuDNN 7.6, GTX 1070 8G

**Describe the current behavior**
tf.keras.layers.ReLU()(x) sets `x._keras_mask` to None.

`tf.keras.layers.Embedding(mask_zero=True)(x)` and `tf.keras.layers.GRU(mask=self_defined_mask)(x)` set a bool Tensor `mask` to `x._keras_mask`. This mask will be passed on to Loss and Metrics and serves as `sample_weight` when calculating the cross_entropy_loss and metrics.accuracy etc. However, it just disappeared after a ReLU layer.

I found tf.keras.layers.ReLU() call K.relu() which in turn calls tf.nn.relu(). `tf.nn.relu()`, a legacy of TF 1.x which doesn't support `_keras_mask`, seems to be the cause of the unintentional drop of the mask.

**Describe the expected behavior**
tf.keras.layers.ReLU()(x) just leaves `x._keras_mask` as it was.

**Code to reproduce the issue** Provide a reproducible test case that is the
bare minimum necessary to generate the problem.

```
class GRU4REC(tf.keras.Model):
    def __init__(self, vocabulary_size, embedding_dim, hidden_units):
        super(GRU4REC, self).__init__()
        self.Embedding = tf.keras.layers.Embedding(vocabulary_size, embedding_dim, mask_zero=True)
        self.FC1 = tf.keras.layers.Dense(hidden_units * 3, tf.nn.relu)
        self.BN1 = tf.keras.layers.BatchNormalization()
        self.RELU = tf.keras.layers.ReLU()
        self.GRU = tf.keras.layers.GRU(hidden_units, return_sequences=True)
        self.FC2 = tf.keras.layers.Dense(vocabulary_size)
        self.BN2 = tf.keras.layers.BatchNormalization()

    def call(self, inputs, training=None, mask=None):
        x = self.Embedding(inputs)
        x = self.FC1(x)
        x = self.BN1(x)
        print(f'before relu: {x._keras_mask}') # print a Tensor
        x = self.RELU(x)
        print(f'after relu: {x._keras_mask}') # print None
        x = self.GRU(x)
        x = self.FC2(x)
        x = self.BN2(x)
        x = self.RELU(x)
        # pred = tf.argmax(x, -1)
        # tf.print('++++++++++')
        # tf.print(tf.shape(pred))
        # eq = tf.equal(inputs, tf.cast(pred, tf.int32))
        # tf.print(tf.reduce_sum(tf.cast(eq, tf.float32)))
        # tf.print('---------')
        return x
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

On TF 2.0

```
2020-02-19 20:24:05.261124: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library nvcuda.dll
2020-02-19 20:24:05.292473: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1555] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce GTX 1070 computeCapability: 6.1
coreClock: 1.7845GHz coreCount: 15 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 238.66GiB/s
2020-02-19 20:24:05.292814: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_101.dll
2020-02-19 20:24:05.425947: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_10.dll
2020-02-19 20:24:05.503908: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_10.dll
2020-02-19 20:24:05.538431: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_10.dll
2020-02-19 20:24:05.660730: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_10.dll
2020-02-19 20:24:05.738706: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_10.dll
2020-02-19 20:24:05.944062: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2020-02-19 20:24:05.945225: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1697] Adding visible gpu devices: 0
2020-02-19 20:24:05.947055: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1555] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce GTX 1070 computeCapability: 6.1
coreClock: 1.7845GHz coreCount: 15 deviceMemorySize: 8.00GiB deviceMemoryBandwidth: 238.66GiB/s
2020-02-19 20:24:05.947607: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_101.dll
2020-02-19 20:24:05.947889: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_10.dll
2020-02-19 20:24:05.948150: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cufft64_10.dll
2020-02-19 20:24:05.948418: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library curand64_10.dll
2020-02-19 20:24:05.948677: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusolver64_10.dll
2020-02-19 20:24:05.948922: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cusparse64_10.dll
2020-02-19 20:24:05.949206: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudnn64_7.dll
2020-02-19 20:24:05.950821: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1697] Adding visible gpu devices: 0
2020-02-19 20:24:06.889483: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1096] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-02-19 20:24:06.889680: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102]      0 
2020-02-19 20:24:06.889787: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] 0:   N 
2020-02-19 20:24:06.891077: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1241] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6376 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1070, pci bus id: 0000:01:00.0, compute capability: 6.1)
before relu: Tensor("gr_u4rec/embedding/NotEqual:0", shape=(None, None), dtype=bool)
after relu: None
before relu: Tensor("model/gr_u4rec/embedding/NotEqual:0", shape=(None, None), dtype=bool)
after relu: None
2020-02-19 20:24:08.567321: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cublas64_10.dll

Process finished with exit code -1
```

When I was training a lots-of-zero-padded sequences, the loss went down smoothly as expected while the accuracy soared dramatically. However, everything went just right when I removed ReLU. Hope fix it or help point it out if I was wrong. Thanks~

**Comment by gadagashwini** [comment]
@Rayarrow, Thanks for reporting this issue.
Can you share the complete code to replicate the reported issue. Thanks!

**Comment by Rayarrow** [comment]
The real dataset and code I used to train is too large and messy to be attached here, so I faked a dataset produced by a random sequence generator but I think it's enough to explain my issue.

This script does an evaluation on the model, and then a training process followed by another round of evaluation. Please run this script twice:
    1. With ReLU. The accuracy will soar (~40%) during the training process and also in the second round of evaluation compared to the first round.
    2. Without ReLU (comment line 21 and line 26). The accuracy just levels out (~0.2%) as the model go through training and evaluation.
    
Besides, object `x` will be printed twice to the console before and after it goes through layer `ReLU`, and it turn into `None` after it went through ReLU. So I think `ReLU` is the cause.

Thanks!

```
import numpy as np
import tensorflow as tf


class GRU4REC(tf.keras.Model):
    def __init__(self, vocabulary_size, embedding_dim, hidden_units):
        super(GRU4REC, self).__init__()
        self.Embedding = tf.keras.layers.Embedding(vocabulary_size, embedding_dim, mask_zero=True)
        self.FC1 = tf.keras.layers.Dense(hidden_units * 3, tf.nn.relu)
        self.BN1 = tf.keras.layers.BatchNormalization()
        self.RELU = tf.keras.layers.ReLU()
        self.GRU = tf.keras.layers.GRU(hidden_units, return_sequences=True)
        self.FC2 = tf.keras.layers.Dense(vocabulary_size)
        self.BN2 = tf.keras.layers.BatchNormalization()

    def call(self, inputs, training=None, mask=None):
        x = self.Embedding(inputs)
        x = self.FC1(x)
        x = self.BN1(x)
        print(f'before relu: {x._keras_mask}')  # print a Tensor
        x = self.RELU(x)
        print(f'after relu: {x._keras_mask}')  # print None
        x = self.GRU(x)
        x = self.FC2(x)
        x = self.BN2(x)
        x = self.RELU(x)
        return x


X_input = tf.keras.layers.Input([None])
encoder = GRU4REC(200, 128, 100)
model = tf.keras.Model(inputs=X_input, outputs=encoder(X_input))

# After compile, GRU4REC.call() will be invoked and the Tensor `x` before and after ReLU will be printed to the console.
model.compile(tf.optimizers.Adam(), tf.losses.SparseCategoricalCrossentropy(),
              metrics=[tf.metrics.SparseCategoricalAccuracy()])


def random_sequence_generator():
    for i in range(1000):
        sequence = np.random.randint(0, 199, np.random.randint(3, 11, 1))
        yield sequence[:-1], sequence[1:]


train_dataset = tf.data.Dataset.from_generator(random_sequence_generator, output_types=(tf.int32, tf.int32)).shuffle(
    1000).padded_batch(2, ((None,), (None,)))

valid_dataset = tf.data.Dataset.from_generator(random_sequence_generator,
                                               output_types=(tf.int32, tf.int32)).padded_batch(2, ((None,), (None,)))

model.evaluate(valid_dataset)
model.fit(train_dataset, epochs=2)
model.evaluate(valid_dataset)
```

**Comment by gadagashwini** [comment]
Was able to reproduce the issue with Tf 2.1.
Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/f813eb95dccaff1af7e9e33dd949a363/untitled398.ipynb). Thanks

**Comment by jvishnuvardhan** [comment]
@Rayarrow Looks like this was resolved in `tf-nightly`. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/d97f138de1da3f8e50646692a37830c2/untitled398.ipynb).

Can you please verify once and let us know whether it was resolved or not. Thanks!

**Comment by Rayarrow** [comment]
Ok. Too busy right now and I will try it a few days later. Wait for my message please.
